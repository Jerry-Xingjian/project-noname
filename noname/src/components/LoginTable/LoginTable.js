import React from 'react';
import './LoginTable.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { Link, useNavigate } from 'react-router-dom';
import { login } from '../../utils/api/user';
import { localGet, localSet } from '../../utils/localStorage';

function LoginTable() {
  const navigate = useNavigate();
  const MAX_ATTEMPTS = 3;
  const accountLockInfo = localGet('accountInfo') || [{}];

  const handleSubmit = async (event) => {
    event.preventDefault();
    const email = document.getElementById('floatingInput').value;
    const password = document.getElementById('floatingPassword').value;
    let lockInfo;
    accountLockInfo.forEach((account) => {
      if (account.email === email) {
        lockInfo = account;
      }
    });
    if (lockInfo == null && email !== '') {
      accountLockInfo.push({
        email,
        loginAttempts: 0,
        lockoutTime: null,
      });
      lockInfo = accountLockInfo[accountLockInfo.length - 1];
    }

    // Check if the account is currently locked out
    if (email !== '' && lockInfo.lockoutTime !== null) {
      const timeRemaining = lockInfo.lockoutTime + 60 * 1000 - Date.now();
      if (timeRemaining > 0) {
        localSet('accountInfo', accountLockInfo);
        // eslint-disable-next-line no-alert
        alert(`Your account is locked out for ${Math.ceil(timeRemaining / 1000)} seconds. Please try again later.`);
        return;
      }
      // reset the lockoutTime variable to null to allow the user to try logging in again
      lockInfo.lockoutTime = null;
      lockInfo.loginAttempts = 0;
    }
    // login process
    try {
      await login(email, password);
      const index = accountLockInfo.indexOf(lockInfo);
      accountLockInfo.splice(index, 1);
      localSet('accountInfo', accountLockInfo);
      navigate('/home');
    } catch (err) {
      // if user submit the wrong password
      if (err.message.includes(409)) {
        lockInfo.loginAttempts += 1;
        localSet('accountInfo', accountLockInfo);
        if (lockInfo.loginAttempts > 2) {
          lockInfo.lockoutTime = Date.now();
          // eslint-disable-next-line no-alert, max-len
          alert('You have exceeded the maximum number of login attempts. Your account will be locked out for 1 minute.');
        } else {
          // eslint-disable-next-line no-alert
          alert(`Invalid username or password. You have ${MAX_ATTEMPTS - lockInfo.loginAttempts} attempts remaining.`);
        }
      } else if (err.message.includes(410)) {
        const index = accountLockInfo.indexOf(lockInfo);
        accountLockInfo.splice(index, 1);
        // eslint-disable-next-line no-alert
        alert('This email is not registeed. Check typo or register an account.');
      } else if (err.message.includes(404)) {
        // eslint-disable-next-line no-alert
        alert('Missing email or password');
      }
    }
  };

  // Set an interval to check every 1000 milliseconds (1 second) if the lockout has expired
  setInterval(() => {
  // Retrieve the lock information for the specified account
    let lockInfo;
    try {
      const email = document.getElementById('floatingInput').value;
      lockInfo = accountLockInfo.filter((account) => account.email === email);
    } catch (err) {
      return;
    }

    // Check if the account is currently locked out
    if (lockInfo.length !== 0) {
      if (lockInfo[0].lockoutTime !== null) {
        // If the account is locked out
        const timeRemaining = lockInfo.lockoutTime + 60 * 1000 - Date.now();
        // If the lock has expired
        if (timeRemaining <= 0) {
          lockInfo.lockoutTime = null;
          lockInfo.loginAttempts = 0;
        }
      }
    }
  }, 1000);

  return (
    <div className="container d-flex flex-column w-25 position-absolute top-50 end-0 translate-middle">
      <form className="d-flex flex-column" onSubmit={handleSubmit}>
        <div className="mb-3">
          <h1 className="text-center">Login</h1>
          <p className="text-center text-black-50">
            Login and Stay connected with
            <br />
            your community
          </p>
        </div>
        <div className="form-floating mb-3">
          <input type="email" className="form-control background-white" id="floatingInput" placeholder="name@example.com" name="email" />
          <label htmlFor="floatingInput">Email address</label>
        </div>
        <div className="form-floating mb-3">
          <input type="password" className="form-control background-white" id="floatingPassword" placeholder="Password" name="password" />
          <label htmlFor="floatingPassword">Password</label>
        </div>
        <button type="submit" className="btn btn-danger text-light mb-3">Login</button>
        {/* <button type="button" className="btn btn-primary text-light mb-3">
        Login with Google</button> */}
      </form>
      <div className="text-center text-black-50 d-flex justify-content-center">
        <p className="my-0 me-1">Need an account?</p>
        <Link className="text-danger" to="/register">
          <p>Register here!</p>
        </Link>
      </div>
    </div>
  );
}

export default LoginTable;
