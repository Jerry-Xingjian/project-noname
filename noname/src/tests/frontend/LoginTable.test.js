import { render, screen, fireEvent, waitFor} from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import {createMemoryHistory} from 'history';
import React from "react";
import {Router} from 'react-router-dom'
import '@testing-library/jest-dom'
import LoginTable from '../../components/LoginTable/LoginTable';
import renderer from 'react-test-renderer';
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

describe("Component test", () => {
  test('certain texts exists in the page', () => {
    const history = createMemoryHistory();
    render(
      <Router location={history.location} navigator={history}>
        <LoginTable />
      </Router>,
    )
    // check whether "Login and Stay connected with" exits in the screen or not
    const text = screen.getByText(/Login and Stay connected with/i)
    expect(text).toBeInTheDocument()
  })
});

describe("Snapshot test", () => {
  test('LoginTable component matches snapshot', () => {
    const history = createMemoryHistory();
    const component = renderer.create(
      <Router location={history.location} navigator={history}>
        <LoginTable />
      </Router>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Navigation test", () => {
  let mock;

  beforeAll(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.reset();
  });

  const default_data = [{ test_data: 20}];

  test("Login navigate to register page", async () => {
    // test for different effects than the page actually changing
    // capture the result of a click event and outputting history.location.pathname correctly showing as /register
    const history = createMemoryHistory();
    const spy = jest.spyOn(history, 'push');
    render(
      <Router location={history.location} navigator={history}>
        <LoginTable />
      </Router>,
    )
    const user = userEvent.setup()
    expect(screen.getByText(/Register here!/i)).toBeInTheDocument()
    await user.click(screen.getByText(/Register here!/i))

    // check that the content changed to the new page
    expect(history.location.pathname).toBe('/register');
    expect(spy).toBeCalled();
  });

//   test("Login submit info and navigate to home page", async () => {
//     const history = createMemoryHistory();
//     // spying on history.push
//     const spy = jest.spyOn(history, 'push');
//     render(
//       <Router location={history.location} navigator={history}>
//         <LoginTable />
//       </Router>,
//     )
//     // test user type email and password event
//     const email = screen.getByLabelText("Email address")
//     const password = screen.getByLabelText("Password");
//     await userEvent.type(email, "xwang@gmail.com");
//     await userEvent.type(password, "123456")
//     expect(email).toHaveValue('xwang@gmail.com')
//     expect(password).toHaveValue('123456')
    
//     // Mock POST request when param are `default_data`, 'user_id' and user input email and password
//     mock.onPost().reply(200, default_data)
//     // waitFor may run the callback function, in this case, fireEvent.click
//     // a number of times until the timeout is reached
//     // keep calling axios until got response
//     await waitFor(() => {
//       fireEvent.click(screen.getByRole('button', { name: 'Login' }));
//     });
//     // test whether the axios test is post request or not
//     expect(mock.onPost().reply(200, default_data).history.post[0].method).toEqual("post")
//     // check that the content changed to the new page
//     expect(spy).toBeCalled();
//     expect(history.location.pathname).toBe('/home');
//   });
});

describe("Event test", () => {
  test("Email and password change test", async () => {
    const history = createMemoryHistory();
    render(
      <Router location={history.location} navigator={history}>
        <LoginTable />
      </Router>
    )
    const email = screen.getByLabelText("Email address");
    const password = screen.getByLabelText("Password");
    // initial email and password is expected to be null
    expect(email).toHaveValue('')
    expect(password).toHaveValue('')
    fireEvent.change(email, {target: {value: 'xwang@gmail.com'}})
    fireEvent.change(password, {target: {value: '123456'}})
    // fireEvent.change will change its contents
    expect(email).toHaveValue('xwang@gmail.com')
    expect(password).toHaveValue('123456')
  });
});

