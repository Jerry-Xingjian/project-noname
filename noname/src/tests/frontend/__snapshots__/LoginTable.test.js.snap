// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test LoginTable component matches snapshot 1`] = `
<div
  className="container d-flex flex-column w-25 position-absolute top-50 end-0 translate-middle"
>
  <form
    className="d-flex flex-column"
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <h1
        className="text-center"
      >
        Login
      </h1>
      <p
        className="text-center text-black-50"
      >
        Login and Stay connected with
        <br />
        your community
      </p>
    </div>
    <div
      className="form-floating mb-3"
    >
      <input
        className="form-control background-white"
        id="floatingInput"
        name="email"
        placeholder="name@example.com"
        type="email"
      />
      <label
        htmlFor="floatingInput"
      >
        Email address
      </label>
    </div>
    <div
      className="form-floating mb-3"
    >
      <input
        className="form-control background-white"
        id="floatingPassword"
        name="password"
        placeholder="Password"
        type="password"
      />
      <label
        htmlFor="floatingPassword"
      >
        Password
      </label>
    </div>
    <button
      className="btn btn-danger text-light mb-3"
      type="submit"
    >
      Login
    </button>
    <button
      className="btn btn-primary text-light mb-3"
      type="button"
    >
      Login with Google
    </button>
  </form>
  <div
    className="text-center text-black-50 d-flex justify-content-center"
  >
    <p
      className="my-0 me-1"
    >
      Need an account?
    </p>
    <a
      className="text-danger"
      href="/register"
      onClick={[Function]}
    >
      <p>
        Register here!
      </p>
    </a>
  </div>
</div>
`;
